ksql-engine/src/test/java/io/confluent/ksql/physical/PhysicalPlanBuilderTest.java

```diff
@@ -739,42 +740,333 @@ public void shouldHaveOptionalFieldsInResultSchema() {
         + "col2 FROM "
         + "test1;";
     final String insertIntoQuery = "INSERT INTO s1 SELECT col0, col1, col2 FROM test1;";
-    givenKafkaTopicExists("test1");
-    final List<QueryMetadata> queryMetadataList = KsqlEngineTestUtil.execute(
-        ksqlEngine, createStream + "\n " +
-        csasQuery + "\n " +
-        insertIntoQuery,
-        ksqlConfig,
-        Collections.emptyMap());
+    givenKafkaTopicsExist("test1");
+    final List<QueryMetadata> queryMetadataList = execute(
+        CREATE_STREAM_TEST1 + csasQuery + insertIntoQuery);
     final Schema resultSchema = queryMetadataList.get(0).getResultSchema();
     resultSchema.fields().forEach(
         field -> Assert.assertTrue(field.schema().isOptional())
     );
-    closeQueries(queryMetadataList);
   }
 
   @Test
   public void shouldSetIsKSQLSinkInMetastoreCorrectly() {
     final String csasQuery = "CREATE STREAM s1 AS SELECT col0, col1, col2 FROM test1;";
     final String ctasQuery = "CREATE TABLE t1 AS SELECT col0, COUNT(*) FROM test1 GROUP BY col0;";
-    givenKafkaTopicExists("test1");
-    KsqlEngineTestUtil.execute(
-        ksqlEngine,
-        createStream + "\n " + csasQuery + "\n " + ctasQuery,
-        ksqlConfig,
-        Collections.emptyMap());
+    givenKafkaTopicsExist("test1");
+    execute(CREATE_STREAM_TEST1 + csasQuery + ctasQuery);
     assertThat(ksqlEngine.getMetaStore().getSource("TEST1").getKsqlTopic().isKsqlSink(), equalTo(false));
     assertThat(ksqlEngine.getMetaStore().getSource("S1").getKsqlTopic().isKsqlSink(), equalTo(true));
     assertThat(ksqlEngine.getMetaStore().getSource("T1").getKsqlTopic().isKsqlSink(), equalTo(true));
   }
 
+  @Test
+  public void shouldRepartitionLeftStreamIfNotCorrectKey() {
+    // Given:
+    givenKafkaTopicsExist("test2", "test3");
+    execute(CREATE_STREAM_TEST2 + CREATE_STREAM_TEST3);
+
+    // When:
+    final QueryMetadata result = execute("CREATE STREAM s1 AS "
+        + "SELECT * FROM test2 JOIN test3 WITHIN 1 SECOND "
+        + "ON test2.col1 = test3.id;")
+        .get(0);
+
+    // Then:
+    assertThat(result.getExecutionPlan(),
+        containsString("[ REKEY ] | Schema: [TEST2.ROWTIME : BIGINT"));
+  }
+
+  @Test
+  public void shouldRepartitionRightStreamIfNotCorrectKey() {
+    // Given:
+    givenKafkaTopicsExist("test2", "test3");
+    execute(CREATE_STREAM_TEST2 + CREATE_STREAM_TEST3);
+
+    // When:
+    final QueryMetadata result = execute("CREATE STREAM s1 AS "
+        + "SELECT * FROM test2 JOIN test3 WITHIN 1 SECOND "
+        + "ON test2.id = test3.col0;")
+        .get(0);
+
+    // Then:
+    assertThat(result.getExecutionPlan(),
+        containsString("[ REKEY ] | Schema: [TEST3.ROWTIME : BIGINT"));
```

Comment:
is there anything in the execution plan that indicates we are rekeying to the correct column?

Reply:
I'll take a look.  The main thing I was looking for was to ensure presence / absence of the repartition step -  but you're right, may as well check _what_ its being reparatitioned by as well.

Reply:
Unfortunately, I can't see anything in the plan / query that currently shows _what_ it's partitioned by.